Delete units

sparakala21 opened the pull request:
# Description
I added the handleDelete function in EditUnitModalComponent.tsx so that admins can delete units when it is necessary.
I also added a way to ensure safety when an admin changes a meters intrinsic properties. 
(Please include a summary of the change and which issue is touched on. Please also include relevant motivation and context.)

Fixes #1020 and Fixes #1042.

## Type of change

- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [ ] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [ ] I have removed text in ( ) from the issue request
- [ ] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.github.io/developer/cla.html) and each author is listed in the Description section.

## Limitations

sparakala21 committed:
added a delete button

sparakala21 committed:
unit usage detection logic

sparakala21 committed:
react alert instead of logging on the console

sparakala21 committed:
translate error messages into french and spanish

sparakala21 committed:
i fixed a dumb bug that wasn't giving me problems

sparakala21 committed:
completely redesigned the function based on Steve

sparakala21 committed:
Merge branch 'deleteUnits' of deleteUnits

sparakala21 committed:
i did not write this code

sparakala21 committed:
cleanup

sparakala21 committed:
small changes

sparakala21 committed:
bug fix

sparakala21 committed:
translations

sparakala21 committed:
the button said Discard changes instead Delete Changes

huss commented on src/client/app/components/groups/EditGroupModalComponent.tsx:
```diff
@@ -132,7 +132,7 @@ export default function EditGroupModalComponent(props: EditGroupModalComponentPr
 	*/
 	const [validGroup, setValidGroup] = useState(false);
-	useEffect(() => {
+	useEffect(() => {deleteConfirmText
```
I'm unclear what "deleteConfirmText" is for here. Could you explain?

sparakala21 replied:
ill be honest i dont think that was me

huss replied:
This is gone with the change to RTK so no longer an issue.

huss commented on src/client/app/components/meters/EditMeterModalComponent.tsx:
```diff
@@ -240,6 +243,22 @@ export default function EditMeterModalComponent(props: EditMeterModalComponentPr
+			for (let i = 0; i < value.deepMeters.length; i++) {
+				console.log(value.deepMeters[i], props.meter.id)
+				if(value.deepMeters[i]== props.meter.id)
+				{
```
OED follows the default VSC formatting where the { is on the same line as the if. There are also spacing issues. Would you please do a "Format Document" to clean this up.

huss replied:
fixed

huss commented on src/client/app/components/meters/EditMeterModalComponent.tsx:
```diff
@@ -240,6 +243,22 @@ export default function EditMeterModalComponent(props: EditMeterModalComponentPr
 				props.meter.maxError != state.maxError ||
 				props.meter.disableChecks != state.disableChecks
 			);
+		let error_message = '';
```
Could some comments be added to describe the basic functioning of this code?

huss commented on src/client/app/components/meters/EditMeterModalComponent.tsx:
```diff
@@ -240,6 +243,22 @@ export default function EditMeterModalComponent(props: EditMeterModalComponentPr
+		let error_message = '';
+		for (const value of Object.values(groups)) {
+			for (let i = 0; i < value.deepMeters.length; i++) {
+				console.log(value.deepMeters[i], props.meter.id)
```
Was this console.log for debugging? In general, we don't have them in the final code.

huss replied:
removed

huss commented on src/client/app/components/meters/EditMeterModalComponent.tsx:
```diff
@@ -240,6 +243,22 @@ export default function EditMeterModalComponent(props: EditMeterModalComponentPr
 				props.meter.disableChecks != state.disableChecks
 			);
+		let error_message = '';
+		for (const value of Object.values(groups)) {
```
I think this check is run every time. Is that correct? For example, if there is no change should it run? Another is if you only change the name then does it matter if it is in another group? I suspect this should be restricted to certain cases. Sorry if the issue and our conversations did not cover this and let me know if you think otherwise.

sparakala21 replied:
i changed the implementation so that it only runs when it needs to. What do you think?

huss replied:
I also restricted it more.

huss commented on src/client/app/components/meters/EditMeterModalComponent.tsx:
```diff
@@ -240,6 +243,22 @@ export default function EditMeterModalComponent(props: EditMeterModalComponentPr
+				}
+			}
+		}
+		state.unitId = props.meter.unitId;
```
First, I'm not sure why the state is always being reset. Second, directly setting state can cause issues. I think using setState() is the way to go here.

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -22,6 +22,7 @@ import { Dispatch } from 'types/redux/actions';
 import { useSelector } from 'react-redux';
 import { State } from 'types/redux/state';
 
```
Please remove new blank line.

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -69,6 +70,45 @@ export default function EditUnitModalComponent(props: EditUnitModalComponentProp
 		setState({ ...state, [e.target.name]: Number(e.target.value) });
 	}
 
+
```
Extra blank line.

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -69,6 +70,45 @@ export default function EditUnitModalComponent(props: EditUnitModalComponentProp
 
+
+
+	// Log the Redux state to the console
```
See formatting comment in edit meter. Other comments on alert, what done when issue found and commenting apply here.

huss replied:
done

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -69,6 +70,45 @@ export default function EditUnitModalComponent(props: EditUnitModalComponentProp
+				+ translate('unit') + ' "' + state.id+ '" as a destination unit\n';
+			}
+		}
+		//finish meter graphicalUnitId and UnitId
```
Are these TODO items? Not sure about these two comments.

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -69,6 +70,45 @@ export default function EditUnitModalComponent(props: EditUnitModalComponentProp
+		}
+
+
+		//
```
Unsure about these comments.

huss commented on src/client/app/components/meters/EditMeterModalComponent.tsx:
```diff
@@ -240,6 +243,22 @@ export default function EditMeterModalComponent(props: EditMeterModalComponentPr
+
+		if (error_message) {
+			error_message = translate('meter.is.not.editable') + '\n' + error_message;
+			alert(error_message);
```
I know alert is used elsewhere but we are trying to convert to Reaxct friendly alerts. Please look at src/client/app/utils/notifications.ts for success and error message functions. I think they would work here.

Also, I'm unclear that this stops the changes. After the alert it does not seem to have logic to stop the following code to update the meter.

sparakala21 replied:
ok i made the change but now it displays both alerts. Im still figuring out how to fix that.

huss replied:
Fixed

huss commented on src/client/app/translations/data.js:
```diff
@@ -334,7 +334,7 @@ const localeData = {
 		"meter.is.not.displayable": "Display Disabled",
 		"meter.is.not.enabled": "Updates Disabled",
-		"meter.previousEnd": "Previous End Time Stamp:",
+				"meter.previousEnd": "Previous End Time Stamp:",
```
extra spacing at start of line

huss commented on src/client/app/translations/data.js:
```diff
@@ -785,6 +786,7 @@ const localeData = {
 		"meter.is.enabled": "Mises à Jour Activées",
 		"meter.is.not.displayable": "Affichage Désactivé",
 		"meter.is.not.enabled": "Mises à Jour Désactivées",
+		"meter.is.not.editable": "(need French) This meter cannot be changed",
```
I don't see this key under en.

huss commented on src/client/app/components/meters/EditMeterModalComponent.tsx:
```diff
@@ -240,6 +243,22 @@ export default function EditMeterModalComponent(props: EditMeterModalComponentPr
+				console.log(value.deepMeters[i], props.meter.id)
+				if(value.deepMeters[i]== props.meter.id)
+				{
+					error_message += translate('group') + value.name + ' uses ' + translate('meter') + ' "' + metersByID[value.deepMeters[i]].name + '" \n';
```
The word 'uses' is not translated. This is true edit unit file too.

huss commented on src/client/app/components/meters/EditMeterModalComponent.tsx:
```diff
@@ -240,6 +243,22 @@ export default function EditMeterModalComponent(props: EditMeterModalComponentPr
+		state.unitId = props.meter.unitId;
+
+		if (error_message) {
+			error_message = translate('meter.is.not.editable') + '\n' + error_message;
```
Maybe use += to be consistent with code above?

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -69,6 +70,45 @@ export default function EditUnitModalComponent(props: EditUnitModalComponentProp
+		let error_message = '';
+		for (const value of Object.values(meterState)) {
+			if(value.unitId==state.id){
+				error_message += translate('meter') + value.name + ' uses ' + translate('unit') + ' "' + state.id+ '" \n';
```
Using the state.id instead of the identifier. Applies below too.

huss reviewed (changes requested):
Thanks to @sparakala21 for more work on OED and this PR. I have made a number of comments. Given this may cause changes in the code, I am going to wait to do testing until they are resolved.

sparakala21 committed:
small fixes

sparakala21 committed:
i didnt use success notification

sparakala21 committed:
some more fixes

sparakala21 committed:
stop using alerts in Edit Unit

sparakala21 committed:
alert->notifyUser

sparakala21 committed:
delete unused import+mistake with groups

huss committed:
formatting; msg format

The newly used error notification does not honor formatting.
A TODO was added and msg made okay without line breaks.

huss committed:
PR 1065

huss commented:
@sparakala21 I could not easily separate out the changes I had tried so I decided to push then out to the PR. There are some comments about open items. Feel free to change my code as needed and to address the items left. Let me know if I can help and thanks for looking into this.

huss commented:
I just wanted to document that this PR is waiting on PR #1113 since it changes how Redux state is accessed. After that, this work will be updated and all concerns addressed.

huss committed:
Merge remote-tracking branch 'origin/development' into pr/sparakala21/1065

huss committed:
fix conversion in unit delete

- Fixes issue with unit state.
- Create function to give arrow for conversion display

huss committed:
add check of unit use as DGU when delete

huss committed:
add delete of unit

Any error is not being shown to admin

huss committed:
fix formatting

huss committed:
show admin delete unit msg

huss committed:
try to remove CodeQL warning

huss committed:
fix variable name

huss committed:
apply fix to for CodeQL to other places

huss committed:
fix up edit meter

Only stops meter unit change if group uses meter.
Also only does check if meter has changes.

huss committed:
format/comment

huss commented:
I updated the PR to merge in #1113 and edited so uses the Redux Toolkit. I also made all the other needed changes to it works as desired. It should be ready to merge.
